### PR TITLE
Additional Docstrings and Comments to `NeuralDual`-Module

### PR DESCRIPTION
**Pull request by bunnech**
Addressing issues [# 110](https://github.com/ott-jax/ott/issues/110) and [#105](https://github.com/ott-jax/ott/pull/105), this is a small pull request adding more docstrings and comments to the existing code 👀.

### CODE REVIEW

File: ott/core/neuraldual.py [hunk 1]
```diff
@@ -30,6 +30,12 @@
 class NeuralDualSolver:
   r"""Solver of the ICNN-based Kantorovich dual.
 
+  Learn the optimal transport between two distributions, denoted source
+  and target, respectively. This is achieved by parameterizing the two
+  Kantorovich potentials, `g` and `f`, by two input convex neural networks.
+  `\nabla g` hereby transports source to target cells, and `\nabla f`
```
Comment (michalk8) [inline ott/core/neuraldual.py:36]: The \`\nabla g\` doesn't render, try :math:\`\nabla g\` instead.

File: ott/core/neuraldual.py [hunk 2]
```diff
@@ -30,6 +30,12 @@
...
+  target to source cells.
+  
   The algorithm is described in:
```
Comment (michalk8) [inline ott/core/neuraldual.py:39]: Would use :cite:\`makkuva:20\` here.

File: ott/core/neuraldual.py [hunk 3]
```diff
@@ -329,6 +335,11 @@ def penalize_weights_icnn(self, params):
 @jax.tree_util.register_pytree_node_class
 class NeuralDual:
   r"""Neural Kantorovich dual.
+  
+  This class contains the solution of the trained Kantorovich neural
+  dual by holding the trained potentials `g` and `f`. `\nabla g`
```
Comment (michalk8) [inline ott/core/neuraldual.py:340]: Same comment regarding math rendering.

File: ott/core/neuraldual.py [hunk 4]
```diff
@@ -375,7 +400,15 @@ def inverse_transport(self, data: jnp.ndarray) -> jnp.ndarray:
...
   def distance(self, source: jnp.ndarray, target: jnp.ndarray) -> float:
-    """Given potentials f and g, compute the overall distance."""
+    r"""Given potentials `f` and `g`, compute the overall distance.
+
+    Args:
+      source: samples of source distribution
+      target: samples of target distribution
+ 
+    Returns:
+      Wasserstein distance $W^2_2$ assuming $|x-y|^2$ as ground distance
```
Comment (michalk8) [inline ott/core/neuraldual.py:410]: The math doesn't render, in rST it should be, e.g., :math:\`W^2_2\`.

**Review by michalk8 — CHANGES_REQUESTED** [review]
Looks good, just minor tweaks like fixing the math rendering and linter.

**Comment by JTT94** [comment]
Minor, I assume best practice is to import the module not the class or function
https://github.com/ott-jax/ott/blob/5657e786c99afb808edd98e061ec316eed0942be/ott/core/icnn.py#L26

**Comment by bunnech** [comment]
Thanks for your comments @michalk8 and James. I hope you are enjoying the summer!
I also updated two lines of `CONTRIBUTING.md` as this seemed to be outdated. 

Let me know if we are ready to merge!

File: ott/core/neuraldual.py [hunk 5]
```diff
@@ -330,6 +333,11 @@ def penalize_weights_icnn(self, params):
 class NeuralDual:
   r"""Neural Kantorovich dual.
 
+  This class contains the solution of the trained Kantorovich neural
+  dual by holding the trained potentials `g` and `f`. :math:`\nabla g`
+  hereby transports source to target cells, and :math:`\nabla f` target
+  to source cells
+
   Attributes:
```
Comment (michalk8) [inline ott/core/neuraldual.py:341]: Should be `Args:`, not `Attributes`. Also please type the variables.

**Review by michalk8 — CHANGES_REQUESTED** [review]
Could you please use :cite:\`bunne:22\` [here](https://github.com/ott-jax/ott/blob/main/ott/core/icnn.py#L37) and fix the typo [here](https://github.com/ott-jax/ott/blob/main/ott/core/icnn.py#L35)?
Maybe even :cite:\`amos:17\` [here](https://github.com/ott-jax/ott/blob/main/ott/core/icnn.py#L16), but I'm just being picky (this doesn't get rendered).

Furthermore, would be nice to:
- type `NeuralDualSolver.{clip_weights_icnn,penalize_weights_icnn}` and document them or make them private (if they are supposed to be used internally only)
- same for `ICNN.compute_{gaussian,identity}_map
- type+document `NeuralDual.{f,g}` properties (also related to the 1 review comment)

After this is done, will be happy to merge.

**Comment by bunnech** [comment]
> Could you please use :cite:`bunne:22` [here](https://github.com/ott-jax/ott/blob/main/ott/core/icnn.py#L37) and fix the typo [here](https://github.com/ott-jax/ott/blob/main/ott/core/icnn.py#L35)?
Maybe even :cite:`amos:17` [here](https://github.com/ott-jax/ott/blob/main/ott/core/icnn.py#L16), but I'm just being picky (this doesn't get rendered).

Done, but added :cite:`amos:17` in the documentation, not the file title.

> Furthermore, would be nice to:
> 
> * type `NeuralDualSolver.{clip_weights_icnn,penalize_weights_icnn}` and document them or make them private (if they are supposed to be used internally only)
> * same for `ICNN.compute_{gaussian,identity}_map

Done, I made them private. Not sure what you want exactly.

File: ott/core/icnn.py [hunk 6]
```diff
@@ -32,9 +32,11 @@
 
 
 class ICNN(nn.Module):
-  """Input convex neural network (ICNN) architeture with initialization.
+  """Input convex neural network (ICNN) architecture with initialization.
 
-  Containing initialization schemes introduced in Bunne+(2022).
+  Implementation of input convex neural networks as introduced in
+  :cite:`amos:17. Containing initialization schemes proposed by
```
Comment (michalk8) [inline ott/core/icnn.py:38]: You're missing the 2nd \` in :cite:\`amos:17\`, causing sphinx to parse the docstring incorrectly as `Implementation of input convex neural networks as introduced in [].`
(note to self - consider treating warnings as errors when building docs).
Furthermore, would rephrase `... . Containing initialization`  to either `, containing initialization schemes` or `... . Contains initialization` (sorry for being picky).
Lastly, the `Bunne+(2022)` below is not really necessary, the :cite:\`bunne:22\` already renders it nicely.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: ott/core/icnn.py (hunk 6 above)
Comment (bunnech) [reply ott/core/icnn.py]: Fixed it! Thanks a lot!

**Review by michalk8 — APPROVED** [review]

**Comment by michalk8** [comment]
LGTM, thanks a lot @bunnech !
